model/labels/labels.go

```diff
@@ -359,39 +383,212 @@ func FromStrings(ss ...string) Labels {
 	if len(ss)%2 != 0 {
 		panic("invalid number of strings")
 	}
-	res := make(Labels, 0, len(ss)/2)
+	ls := make([]Label, 0, len(ss)/2)
 	for i := 0; i < len(ss); i += 2 {
-		res = append(res, Label{Name: ss[i], Value: ss[i+1]})
+		ls = append(ls, Label{Name: ss[i], Value: ss[i+1]})
 	}
 
-	sort.Sort(res)
-	return res
+	sort.Sort(labelSlice(ls))
+	return New(ls...)
 }
 
 // Compare compares the two label sets.
 // The result will be 0 if a==b, <0 if a < b, and >0 if a > b.
+// TODO: replace with Less function - Compare is never needed.
+// TODO: just compare the underlying strings when we don't need alphanumeric sorting.
 func Compare(a, b Labels) int {
-	l := len(a)
-	if len(b) < l {
-		l = len(b)
-	}
-
-	for i := 0; i < l; i++ {
-		if a[i].Name != b[i].Name {
-			if a[i].Name < b[i].Name {
+	l := len(a.data)
+	if len(b.data) < l {
+		l = len(b.data)
+	}
+
+	ia, ib := 0, 0
+	for ia < l {
+		var aName, bName string
+		aName, ia = decodeString(a.data, ia)
+		bName, ib = decodeString(b.data, ib)
+		if aName != bName {
+			if aName < bName {
 				return -1
 			}
 			return 1
 		}
-		if a[i].Value != b[i].Value {
-			if a[i].Value < b[i].Value {
+		var aValue, bValue string
+		aValue, ia = decodeString(a.data, ia)
+		bValue, ib = decodeString(b.data, ib)
+		if aValue != bValue {
+			if aValue < bValue {
 				return -1
 			}
 			return 1
 		}
 	}
 	// If all labels so far were in common, the set with fewer labels comes first.
-	return len(a) - len(b)
+	return len(a.data) - len(b.data)
+}
+
+// Copy labels from b on top of whatever was in ls previously, reusing memory or expanding if needed.
+func (ls *Labels) CopyFrom(b Labels) {
+	ls.data = b.data // strings are immutable
+}
+
+func EmptyLabels() Labels {
+	return Labels{}
+}
+
+// IsEmpty returns true if ls represents an empty set of labels.
+func (ls Labels) IsEmpty() bool {
+	return len(ls.data) == 0
+}
+
+// Len returns the number of labels; it is relatively slow.
+func (ls Labels) Len() int {
```

Comment:
Depending on how much it is needed it might be worth to track the len in a separate field.

Reply:
We can iterate over it too in later PRs (: